clippy_lints/src/while_pop_unwrap.rs

```diff
@@ -0,0 +1,122 @@
+use clippy_utils::{diagnostics::span_lint_and_then, match_def_path, paths, source::snippet};
+use rustc_errors::Applicability;
+use rustc_hir::{Expr, ExprKind, PatKind, Stmt, StmtKind, UnOp};
+use rustc_lint::LateContext;
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+use rustc_span::{symbol::Ident, Span};
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Looks for loops that check for emptiness of a `Vec` in the condition and pop an element
+    /// in the body as a separate operation.
+    ///
+    /// ### Why is this bad?
+    /// Such loops can be written in a more idiomatic way by using a while..let loop and directly
+    /// pattern matching on the return value of `Vec::pop()`.
+    ///
+    /// ### Example
+    /// ```rust
+    /// let mut numbers = vec![1, 2, 3, 4, 5];
+    /// while !numbers.is_empty() {
+    ///     let number = numbers.pop().unwrap();
+    ///     // use `number`
+    /// }
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// let mut numbers = vec![1, 2, 3, 4, 5];
+    /// while let Some(number) = numbers.pop() {
+    ///     // use `number`
+    /// }
+    /// ```
+    #[clippy::version = "1.70.0"]
+    pub WHILE_POP_UNWRAP,
+    style,
+    "checking for emptiness of a `Vec` in the loop condition and popping an element in the body"
+}
+declare_lint_pass!(WhilePopUnwrap => [WHILE_POP_UNWRAP]);
+
+fn report_lint(cx: &LateContext<'_>, pop_span: Span, ident: Option<Ident>, loop_span: Span, receiver_span: Span) {
+    span_lint_and_then(
+        cx,
+        WHILE_POP_UNWRAP,
+        pop_span,
+        "you seem to be trying to pop elements from a `Vec` in a loop",
+        |diag| {
+            diag.span_suggestion(
```

Comment:
Usually, suggestions should contain all changes. To do this right you could use [`clippy_utils::multispan_sugg`](https://doc.rust-lang.org/nightly/nightly-rustc/clippy_utils/diagnostics/fn.multispan_sugg.html), use the `while let Some(..)` message for the `while`loop header and an empty `String` for the pop+unwrapping statement.

There is also a `_with_applicability` variant you could use to make the lint auto-applicable if the snippets are from valid Rust code.

That said, as a style lint, it should probably not apply to macro-generated code. You can check `expr.span.from_expansion()` to rule that out.

Reply:
> That said, as a style lint, it should probably not apply to macro-generated code. You can check expr.span.from_expansion() to rule that out.

The `check()` function is called by the `Loops` pass, which already checks if this is macro-expanded code if I understand it right.

https://github.com/rust-lang/rust-clippy/blob/203c909d04ffc0b2c22f97cb304649aac237087e/clippy_lints/src/loops/mod.rs#L622-L625